File: src/toolchains/npm.py
import subprocess
import json
import shutil


def scan():
    result = {
        "tool": "npm",
        "available": False,
        "node_version": None,
        "npm_version": None,
        "outdated_global": [],
        "issues": [],
        "status": "ok",
    }

    if not shutil.which("node"):
        result["status"] = "skipped"
        result["issues"].append("Node.js not installed")
        return result

    try:
        node_v = subprocess.run(
            ["node", "--version"], capture_output=True, text=True, timeout=5
        )
        result["node_version"] = node_v.stdout.strip()

        if shutil.which("npm"):
            npm_v = subprocess.run(
                ["npm", "--version"], capture_output=True, text=True, timeout=5
            )
            result["npm_version"] = npm_v.stdout.strip()

            outdated = subprocess.run(
                ["npm", "outdated", "-g", "--json"],
                capture_output=True,
                text=True,
                timeout=30,
            )
            if outdated.returncode == 0 and outdated.stdout.strip():
                data = json.loads(outdated.stdout)
                result["outdated_global"] = [
                    {
                        "name": pkg,
                        "current": info.get("current", "?"),
                        "latest": info.get("latest", "?"),
                    }
                    for pkg, info in data.items()
                ]

        result["status"] = "ok" if not result["outdated_global"] else "warning"
        if result["outdated_global"]:
            result["issues"].append(
                f"{len(result['outdated_global'])} outdated global package(s)"
            )
    except Exception as e:
        result["status"] = "error"
        result["issues"].append(str(e))

    return result

File: src/toolchains/npm.rs
use super::*;

pub async fn scan() -> ScanResult {
    if !which("node") {
        return ScanResult::skipped("Node.js not installed");
    }

    let mut result = ScanResult::new("npm");

    if let Ok(ver) = run_cmd("node", &["--version"], None).await {
        result.node_version = Some(ver);
    }

    if !which("npm") {
        return result;
    }

    if let Ok(ver) = run_cmd("npm", &["--version"], None).await {
        result.version = Some(ver);
    }

    if let Ok(out) = run_cmd("npm", &["outdated", "-g", "--json"], None).await {
        result.outdated_global = parse_outdated(&out);
    }

    result.status = if result.outdated_global.is_empty() {
        "ok".into()
    } else {
        "warning".into()
    };

    result
}

pub fn parse_outdated(out: &str) -> Vec<PackageInfo> {
    let mut packages = Vec::new();
    if let Ok(data) = serde_json::from_str::<serde_json::Value>(out) {
        if let Some(obj) = data.as_object() {
            for (name, info) in obj {
                packages.push(PackageInfo {
                    name: name.clone(),
                    current: info["current"].as_str().unwrap_or("?").to_string(),
                    latest: info["latest"].as_str().unwrap_or("?").to_string(),
                });
            }
        }
    }
    packages
}
